File: training/env/lane.py
import numpy as np
from enum import IntEnum
from .generation_config import CONFIG

GRID_W        = 13
MAX_SPEED     = 0.5   # déplacement max en cellules/step (RL)
CELLS_PER_SEC = 3.0   # vitesse visuelle maximale en cases/seconde

class LaneType(IntEnum):
    SAFE  = 0
    GRASS = 1
    ROAD  = 2
    WATER = 3
    LILY  = 4

_TYPE_TO_OBS = {
    LaneType.SAFE:  -1.00,
    LaneType.GRASS: -0.50,
    LaneType.ROAD:   0.00,
    LaneType.WATER:  0.50,
    LaneType.LILY:   1.00,
}

class Lane:
    """
    speed_cps : vitesse signée en cases/seconde (visuel).
                Positif = droite, négatif = gauche.
    score     : score courant du joueur.

    WATER et ROAD utilisent un flux continu d'objets indépendants :
    chaque objet a sa propre largeur (log_size / car_size) et chaque
    espacement est tiré indépendamment (log_space / car_space).
    Les objets entrent d'un côté de l'écran et sortent de l'autre
    sans jamais se répéter.
    """

    def __init__(self, lane_type: LaneType, speed_cps: float,
                 rng: np.random.Generator, score: float = 0):
        self.lane_type = lane_type

        self._speed = float(np.clip(
            (speed_cps / CELLS_PER_SEC) * MAX_SPEED,
            -MAX_SPEED, MAX_SPEED,
        ))

        self._positions: list[float] = []
        self._widths:    list[int]   = []

        # --- HERBE ---
        if lane_type == LaneType.GRASS:
            n_trees = min(CONFIG.tree_count.sample(score, rng), GRID_W)
            if n_trees > 0:
                xs = sorted(rng.choice(GRID_W, size=n_trees, replace=False).tolist())
                for x in xs:
                    self._positions.append(float(x))
                    self._widths.append(1)

        # --- NÉNUPHARS ---
        elif lane_type == LaneType.LILY:
            n_pads = min(CONFIG.lily_count.sample(score, rng), GRID_W)
            if n_pads > 0:
                xs = sorted(rng.choice(GRID_W, size=n_pads, replace=False).tolist())
                for x in xs:
                    self._positions.append(float(x))
                    self._widths.append(1)

        # --- BÛCHES et VOITURES : flux continu ---
        elif lane_type in (LaneType.WATER, LaneType.ROAD):
            self._rng   = rng
            self._score = score
            self._generate_stream()

    # --- génération du flux ---------------------------------------------------

    def _sample_gap(self) -> float:
        if self.lane_type == LaneType.WATER:
            return float(CONFIG.log_space.sample(self._score, self._rng))
        return float(CONFIG.car_space.sample(self._score, self._rng))

    def _sample_width(self) -> int:
        if self.lane_type == LaneType.WATER:
            return CONFIG.log_size.sample(self._score, self._rng)
        return CONFIG.car_size.sample(self._score, self._rng)

    def _generate_stream(self) -> None:
        """Remplit le flux initial : objets de -GRID_W à 2*GRID_W."""
        x = -float(GRID_W) + self._sample_gap()
        while x < 2 * GRID_W:
            width = self._sample_width()
            self._positions.append(float(x))
            self._widths.append(width)
            x += width + self._sample_gap()

    def _trim_and_fill(self, moving_right: bool) -> None:
        """Supprime les objets sortis de l'écran et génère les suivants
        côté entrant pour maintenir un buffer hors-écran d'un GRID_W."""
        if moving_right:
            # sortis à droite
            while self._positions and self._positions[-1] >= GRID_W:
                self._positions.pop()
                self._widths.pop()
            # entrent par la gauche
            while not self._positions or self._positions[0] > -GRID_W:
                gap   = self._sample_gap()
                width = self._sample_width()
                left  = (self._positions[0] - gap - width) if self._positions \
                        else -float(GRID_W)
                self._positions.insert(0, float(left))
                self._widths.insert(0, width)
        else:
            # sortis à gauche
            while self._positions and self._positions[0] + self._widths[0] <= 0:
                self._positions.pop(0)
                self._widths.pop(0)
            # entrent par la droite
            while not self._positions or \
                    self._positions[-1] + self._widths[-1] < 2 * GRID_W:
                gap   = self._sample_gap()
                width = self._sample_width()
                right = (self._positions[-1] + self._widths[-1]) if self._positions \
                        else float(GRID_W)
                self._positions.append(float(right + gap))
                self._widths.append(width)

    # --- propriétés observation RL -------------------------------------------

    @property
    def type(self) -> float:
        return _TYPE_TO_OBS[self.lane_type]

    @property
    def speed(self) -> float:
        return self._speed / MAX_SPEED

    # --- tests de collision --------------------------------------------------
    # Les positions sont désormais absolues (pas de % GRID_W).
    # Un objet à pos avec largeur width occupe [pos, pos+width).

    def has_obstacle_at(self, x: int) -> bool:
        for pos, width in zip(self._positions, self._widths):
            if pos < x + 1 and pos + width > x:
                return True
        return False

    def get_log_at(self, x_float: float):
        """Retourne (pos, width) du log sous le joueur, ou None."""
        hb = 0.25
        for pos, width in zip(self._positions, self._widths):
            if pos < x_float + hb and pos + width > x_float - hb:
                return (pos, width)
        return None

    def is_on_log(self, x_float: float) -> bool:
        hb = 0.25
        for pos, width in zip(self._positions, self._widths):
            if pos < x_float + hb and pos + width > x_float - hb:
                return True
        return False

    def overlaps_cell(self, x: int, hitbox: float = 0.5) -> bool:
        margin   = (1.0 - hitbox) / 2.0
        hb_start = x + margin
        hb_end   = x + 1.0 - margin
        for pos, width in zip(self._positions, self._widths):
            if pos < hb_end and pos + width > hb_start:
                return True
        return False

    def iter_obstacles(self):
        yield from zip(self._positions, self._widths)

    # --- mise à jour des positions -------------------------------------------

    def update(self):
        """Avance les obstacles d'un step RL."""
        self._positions = [p + self._speed for p in self._positions]
        if self.lane_type in (LaneType.WATER, LaneType.ROAD):
            self._trim_and_fill(self._speed > 0)

    def update_visual(self, dt: float):
        """Avance les obstacles en temps réel pour play.py."""
        if not self._positions or self._speed == 0.0:
            return
        delta = (self._speed / MAX_SPEED) * CELLS_PER_SEC * dt
        self._positions = [p + delta for p in self._positions]
        if self.lane_type in (LaneType.WATER, LaneType.ROAD):
            self._trim_and_fill(delta > 0)

File: web/src/game/lane.js
import { GRID_W, PLAYABLE_MIN, PLAYABLE_MAX, PLAYABLE_W, CELLS_PER_SEC, MAX_SPEED } from './constants.js';
import { CONFIG } from './generationConfig.js';

export const LaneType = Object.freeze({ SAFE: 0, GRASS: 1, ROAD: 2, WATER: 3, LILY: 4 });

const TYPE_TO_OBS = {
  [LaneType.SAFE]:  -1.00,
  [LaneType.GRASS]: -0.50,
  [LaneType.ROAD]:   0.00,
  [LaneType.WATER]:  0.50,
  [LaneType.LILY]:   1.00,
};

export class Lane {
  constructor(laneType, speedCps, score = 0) {
    this.laneType   = laneType;
    this._speed     = Math.max(-MAX_SPEED, Math.min(MAX_SPEED, (speedCps / CELLS_PER_SEC) * MAX_SPEED));
    this._score     = score;
    this._positions = [];
    this._widths    = [];

    if (laneType === LaneType.GRASS) {
      // Colonnes hors zone jouable : toujours des arbres (-2 à 14)
      for (let x = -2; x < GRID_W + 2; x++) {
        if (x < PLAYABLE_MIN || x > PLAYABLE_MAX) {
          this._positions.push(x);
          this._widths.push(1);
        }
      }
      // Colonnes jouables : arbres aléatoires
      const n = Math.min(CONFIG.tree_count.sample(score), PLAYABLE_W);
      if (n > 0) {
        for (const s of _sampleNoReplace(PLAYABLE_W, n).sort((a, b) => a - b)) {
          this._positions.push(PLAYABLE_MIN + s);
          this._widths.push(1);
        }
      }
      this._positions.sort((a, b) => a - b);
    } else if (laneType === LaneType.LILY) {
      const n = Math.min(CONFIG.lily_count.sample(score), PLAYABLE_W);
      if (n > 0) {
        for (const s of _sampleNoReplace(PLAYABLE_W, n).sort((a, b) => a - b)) {
          this._positions.push(PLAYABLE_MIN + s);
          this._widths.push(1);
        }
      }
    } else if (laneType === LaneType.WATER || laneType === LaneType.ROAD) {
      this._generateStream();
    }
  }

  // ── propriétés observation RL ─────────────────────────────────────────────
  get type()  { return TYPE_TO_OBS[this.laneType]; }
  get speed() { return this._speed / MAX_SPEED; }

  // ── génération du flux ────────────────────────────────────────────────────
  _sampleGap()   { return this.laneType === LaneType.WATER ? CONFIG.log_space.sample(this._score) : CONFIG.car_space.sample(this._score); }
  _sampleWidth() { return this.laneType === LaneType.WATER ? CONFIG.log_size.sample(this._score)  : CONFIG.car_size.sample(this._score); }

  _generateStream() {
    let x = -GRID_W + this._sampleGap();
    while (x < 2 * GRID_W) {
      const w = this._sampleWidth();
      this._positions.push(x);
      this._widths.push(w);
      x += w + this._sampleGap();
    }
  }

  _trimAndFill(movingRight) {
    if (movingRight) {
      while (this._positions.length && this._positions[this._positions.length - 1] >= GRID_W) {
        this._positions.pop(); this._widths.pop();
      }
      while (!this._positions.length || this._positions[0] > -GRID_W) {
        const gap = this._sampleGap(), width = this._sampleWidth();
        const left = this._positions.length ? this._positions[0] - gap - width : -GRID_W;
        this._positions.unshift(left); this._widths.unshift(width);
      }
    } else {
      while (this._positions.length && this._positions[0] + this._widths[0] <= 0) {
        this._positions.shift(); this._widths.shift();
      }
      while (!this._positions.length ||
             this._positions[this._positions.length - 1] + this._widths[this._widths.length - 1] < 2 * GRID_W) {
        const gap = this._sampleGap(), width = this._sampleWidth();
        const n = this._positions.length;
        const right = n ? this._positions[n - 1] + this._widths[n - 1] : GRID_W;
        this._positions.push(right + gap); this._widths.push(width);
      }
    }
  }

  // ── tests de collision ────────────────────────────────────────────────────
  hasObstacleAt(x) {
    for (let i = 0; i < this._positions.length; i++) {
      if (this._positions[i] < x + 1 && this._positions[i] + this._widths[i] > x) return true;
    }
    return false;
  }

  obstacleAt(x) {
    let cov = 0;
    for (let i = 0; i < this._positions.length; i++) {
      const ov = Math.min(this._positions[i] + this._widths[i], x + 1) - Math.max(this._positions[i], x);
      if (ov > 0) cov += ov;
    }
    return Math.min(cov, 1.0);
  }

  getLogAt(xf) {
    const hb = 0.25;
    for (let i = 0; i < this._positions.length; i++) {
      const p = this._positions[i], w = this._widths[i];
      if (p < xf + hb && p + w > xf - hb) return [p, w];
    }
    return null;
  }

  isOnLog(xf) {
    const hb = 0.25;
    for (let i = 0; i < this._positions.length; i++) {
      if (this._positions[i] < xf + hb && this._positions[i] + this._widths[i] > xf - hb) return true;
    }
    return false;
  }

  overlapsCell(x, hitbox = 0.5) {
    const margin = (1 - hitbox) / 2;
    const hbStart = x + margin, hbEnd = x + 1 - margin;
    for (let i = 0; i < this._positions.length; i++) {
      if (this._positions[i] < hbEnd && this._positions[i] + this._widths[i] > hbStart) return true;
    }
    return false;
  }

  *iterObstacles() {
    for (let i = 0; i < this._positions.length; i++) yield [this._positions[i], this._widths[i]];
  }

  // ── mise à jour ───────────────────────────────────────────────────────────
  update() {
    this._positions = this._positions.map(p => p + this._speed);
    if (this.laneType === LaneType.WATER || this.laneType === LaneType.ROAD)
      this._trimAndFill(this._speed > 0);
  }

  updateVisual(dt) {
    if (!this._positions.length || this._speed === 0) return;
    const delta = (this._speed / MAX_SPEED) * CELLS_PER_SEC * dt;
    this._positions = this._positions.map(p => p + delta);
    if (this.laneType === LaneType.WATER || this.laneType === LaneType.ROAD)
      this._trimAndFill(delta > 0);
  }
}

function _sampleNoReplace(n, k) {
  const arr = Array.from({ length: n }, (_, i) => i);
  for (let i = arr.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [arr[i], arr[j]] = [arr[j], arr[i]];
  }
  return arr.slice(0, k);
}
